acquia/mc-cs-plugin-custom-objects
# MAUT-3700: Deleting custom objects

## Important:

This task consists of 3 PRs:
1. https://github.com/mautic-inc/plugin-custom-objects/pull/228 (this PR)
2. https://github.com/mautic-inc/mautic-cloud/pull/1042
3. https://github.com/mautic-inc/mautic-cloud/pull/1043
They should be merged in the same order. That's important.
The 2nd and the 3rd PR should be merged after we get a confirmation from DevOps that customobjects tube is configured and this PR is merged.
Not for SE and QE: you should only test this 1st and the 3rd PRs.
The 3rd PR has all the changes that are included into the second PR, but it also contains an event subscriber that schedules beanstalk jobs.

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | N
| New feature? | Y
| Automated tests included? | Y
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | [MAUT-3700](https://backlog.acquia.com/browse/MAUT-3700)
| BC breaks? | -
| Deprecations? | -

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
Master objects couldn't be deleted if they had child custom objects.

[//]: # ( As applicable: )
#### Steps to reproduce the bug:
1. Create a parent custom object.
2. Create a child custom object and associate it with your parent custom object.
3. Try to delete your parent custom object. You will get an error (if you instance is not in the debug mode open your browser's developer console and check the status of the AJAX request that is sent when you delete the custom object).

#### Steps to test this PR:
Note for SEs: Checkout the first and the third PR if you test this locally (see list of PRs above). No actions is needed if you test this task in beta environment.
Ignore Jenkins errors in https://github.com/mautic-inc/mautic-cloud/pull/1043. They are caused by missing files in the CustomObjects plugin (deploy issue, will be fixed later).
1. Checkout the 1st and the 3rd PRs as metnioned above. Run `php bin/console mautic:plugins:reload` to apply the migration.
2. Create a parent custom object. Add some custom field to it ( a simple text field is enough).
3. Create some custom items for it (2 is enough).
4. Create a child custom object and associate it with your parent custom object.
5. Associate some custom item with any contact.
6. Delete your parent custom object. You should be able to delete it. You will get a notification that your custom object is being deleted. Custom object must be deleted soon (usually 5-10 seconds is enough). We perform deletion in a separate job so make sure your WorkerConsole.php script is running. Check the console and make sure that the `allyde:customobjects:deletecustomobject` command was executed without errors.
7. Check `custom_item`, `custom_item_xref_contact`, and `custom_item_xref_custom_item` tables. There should be no custom items related to your parent or child custom objects.
8. Go to the contact details page and make sure there is no error.
9. Repeat steps 2-4.
10. Create a segment. Add a segment filter based on any custom field that you created for your custom object.
11. Save the segment.
12. Try to delete a master custom object. You should get an error message saying that your custom object cannot be deleted because it's used in your segment.
13. Check that the custom object wasn't deleted.
14. Remove the custom object based segment filter and make sure you can delete the custom object now.

#### Other areas of Mautic that may be affected by the change:
No

#### List deprecations along with the new alternative:
No

#### List backwards compatibility breaks:
1. There were some BC breaks but both AllydeBundle and CustomObjects are not open source so these BC breaks shouldn't really affect anything.
